### PR TITLE
Add convert read-write-splitting distsql and test case

### LINKED ISSUE #17939: [GSOC 2022] Develop an external tool to convert YAML configuration DistSQL script
This issue is created to better track my PRs for GSoC 2022.
You can find more details about my project [here](https://summerofcode.withgoogle.com/programs/2022/projects/4PnSweAf)

## Background
Apache ShardingSphere is positioned as a Database Plus, and aims at building a standard layer and ecosystem above heterogeneous databases. It focuses on how to reuse existing databases and their respective upper layer, rather than creating a new database. The goal is to minimize or eliminate the challenges caused by underlying databases fragmentation.

Below is my completion plan 

step 1: Parsing YAML files.
Step2: Figure out what we need to complete the output DistSQL RDL  is 'ADD' and 'CREATE' syntax from parsing YAML files.
Step3: Wrap a command line tool to complete the conversion of YAML to DistSQL without starting a proxy.

- [x] ADD RESOURCE dataSource
- [x] CREATE SHARDING TABLE RULE 
- [x] Sharding Binding Table Rule
- [x] Sharding Broadcast Table Rule
- [x] Sharding Scaling Rule
- [x] CREATE READWRITE_SPLITTING RULE
- [x] CREATE DB_DISCOVERY RULE
- [x] CREATE ENCRYPT RULE
- [x] CREATE SHADOW RULE

### PR DESCRIPTION
**Pull request by isHuangXin**
Fixes #17939 

Changes proposed in this pull request:
- Append generateDistSQL by feature SPI
- Add convert read-write-splitting distsql and test case

Need to do:
- Append Dynamic READ-WRITTING RULES by feature SPI

### CODE REVIEW

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java [hunk 1]
```diff
@@ -279,4 +317,79 @@ private String getBindings(final Iterator<String> iterator) {
         result.append(DistSQLScriptConstants.SEMI).append(System.lineSeparator());
         return result.toString();
     }
+    
+    private void appendReadWriteSplittingRules(final Collection<YamlRuleConfiguration> rules, final StringBuilder stringBuilder) {
+        if (rules.isEmpty()) {
+            return;
+        }
+        stringBuilder.append(String.format(DistSQLScriptConstants.CREATE_READWRITE_SPLITTING_RULE));
```
Comment (TeslaCN) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:325]: Why use format here?

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java [hunk 2]
```diff
@@ -279,4 +317,79 @@ private String getBindings(final Iterator<String> iterator) {
...
+            if (iterator.hasNext()) {
+                result.append(DistSQLScriptConstants.COMMA);
+            }
+        }
+        return result.toString();
+    }
+    
+    private String appendLoadBalancer(final String loadBalancerName, final Entry<String, YamlAlgorithmConfiguration> loadBalancers) {
+        StringBuilder result = new StringBuilder();
+        String loadBalancerProperties = new String();
```
Comment (TeslaCN) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:366]: Consider replacing `new String()` with `""`.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java (hunk 1 above)
Comment (isHuangXin) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:325]: fixed

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java (hunk 2 above)
Comment (isHuangXin) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:366]: fixed~

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java [hunk 3]
```diff
@@ -279,4 +317,79 @@ private String getBindings(final Iterator<String> iterator) {
...
+            String loadBalancerType = appendLoadBalancer(entryDataSources.getValue().getLoadBalancerName(), entryLoadBalances);
+            stringBuilder.append(String.format(DistSQLScriptConstants.STATIC_READWRITE_SPLITTING, dataSourceName, writeDataSourceName, readDataSourceNames, loadBalancerType));
+            if (dataSources.hasNext()) {
+                stringBuilder.append(DistSQLScriptConstants.COMMA);
+            }
+        }
+        stringBuilder.append(DistSQLScriptConstants.SEMI).append(System.lineSeparator());
+    }
+    
+    private String appendReadDataSourceNames(final ArrayList<String> readDataSourceNames) {
```
Comment (TeslaCN) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:351]: Consider replacing Collection implementation with interface.

Comment (isHuangXin) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:351]: feel confused

Comment (isHuangXin) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java]: fixed

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java [hunk 4]
```diff
@@ -81,11 +85,44 @@ protected Collection<LocalDataQueryResultRow> getRows(final ContextManager conte
     }
     
     private String generateDistSQL(final YamlProxyDatabaseConfiguration yamlConfig) {
+        StringBuilder result = new StringBuilder();
+        String databaseType = yamlConfig.getDatabaseName();
+        switch (databaseType) {
+            case DistSQLScriptConstants.RESOURCE_DB:
+                result.append(addResourceDistSQL(yamlConfig));
```
[the remark this answers is not in the document]
Comment (isHuangXin) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:92]: get it~

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java [hunk 5]
```diff
@@ -279,4 +310,79 @@ private String getBindings(final Iterator<String> iterator) {
...
+        StringBuilder result = new StringBuilder();
+        String loadBalancerProperties = "";
+        if (loadBalancers.getValue().getProps().isEmpty()) {
+            result.append(String.format(DistSQLScriptConstants.TYPE, loadBalancers.getValue().getType()));
+        } else {
+            Iterator<Entry<Object, Object>> iterator = loadBalancers.getValue().getProps().entrySet().iterator();
+            while (iterator.hasNext()) {
+                Entry<Object, Object> entry = iterator.next();
+                if (loadBalancerName == entry.getKey()) {
+
```
[the remark this answers is not in the document]
Comment (isHuangXin) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:367]: fixed

[the remark this answers is not in the document]
Comment (isHuangXin) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:367]: Are there any other errors ?

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java [hunk 6]
```diff
@@ -279,4 +310,79 @@ private String getBindings(final Iterator<String> iterator) {
...
+            return;
+        }
+        result.append(DistSQLScriptConstants.CREATE_READWRITE_SPLITTING_RULE);
+        for (YamlRuleConfiguration rule : rules) {
+            appendStaticReadWriteSplittingRule(rule, result);
+            // TODO Dynamic READ-WRITE-SPLITTING RULES
+        }
+    }
+    
+    private void appendStaticReadWriteSplittingRule(final YamlRuleConfiguration rule, final StringBuilder result) {
```
Comment (yx9o) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:325]: `ruleConfig` could be better.

Comment (isHuangXin) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/distsql/ral/queryable/ConvertYamlConfigurationHandler.java:325]: fixed

**Review by yx9o — APPROVED** [review]

**Pull request merged**